src/haxeLanguageServer/features/ExtractFunctionFeature.hx

```diff
@@ -0,0 +1,274 @@
+			var path:Path = new Path(fsPath.toString());
+			if (path.ext != "hx")
+				return [];
+
+			if ((doc.tokens == null) || (doc.tokens.tree == null))
+				return [];
+
+			var text:String = doc.getText(params.range);
+			var leftOffset:Int = text.length - text.ltrim().length;
+			var rightOffset:Int = text.length - text.rtrim().length;
+			text = text.trim();
+
+			var tokenStart:Null<TokenTree> = doc.tokens.getTokenAtOffset(doc.offsetAt(params.range.start) + leftOffset);
+			var tokenEnd:Null<TokenTree> = doc.tokens.getTokenAtOffset(doc.offsetAt(params.range.end) - rightOffset);
+			if ((tokenStart == null) || (tokenEnd == null))
+				return [];
+			if (tokenStart.index == tokenEnd.index)
+				return [];
+			// TODO is a minimum of 10 tokens between start and end enough / too much? is there a better solution
+			if (tokenStart.index + 10 > tokenEnd.index)
+				return [];
+
+			var parentOfStart:Null<TokenTree> = findParentFunction(tokenStart);
+			var parentOfEnd:Null<TokenTree> = findParentFunction(tokenEnd);
+			if ((parentOfStart == null) || (parentOfEnd == null))
+				return [];
+			if (parentOfStart.index != parentOfEnd.index)
+				return [];
+			var lastToken:TokenTree = TokenTreeCheckUtils.getLastToken(parentOfStart);
+
+			var rangeIdents:Array<String> = [];
+			var varTokens:Array<TokenTree> = [];
+			var hasReturn:Bool = false;
+			parentOfStart.filterCallback(function(token:TokenTree, index:Int):FilterResult {
+				if (token.index > lastToken.index)
+					return SKIP_SUBTREE;
+				switch (token.tok) {
+					case Const(CIdent(s)):
+						if ((token.index >= tokenStart.index) && (token.index <= tokenEnd.index) && (rangeIdents.indexOf(s) < 0))
+							rangeIdents.push(s);
+					case Dollar(s):
+						if ((token.index >= tokenStart.index) && (token.index <= tokenEnd.index) && (rangeIdents.indexOf(s) < 0))
+							rangeIdents.push("$" + s);
+					case Kwd(KwdReturn):
+						if ((token.index >= tokenStart.index) && (token.index <= tokenEnd.index))
+							hasReturn = true;
+					case Kwd(KwdVar):
+						if (token.index >= tokenStart.index)
+							return GO_DEEPER;
+						if ((token.index >= parentOfStart.index) && (token.index <= lastToken.index))
+							varTokens.push(token);
+					default:
+				}
+				return GO_DEEPER;
+			});
+
+			var returnSpec:String = "";
+			if (hasReturn) {
+				returnSpec = makeReturnSpec(parentOfStart);
+			}
+
+			var newParams:Array<NewFunctionParameter> = copyParentFunctionParameters(parentOfStart, text, rangeIdents);
+			newParams = newParams.concat(localVarsToParameter(varTokens, text, rangeIdents));
+
+			var action:Null<CodeAction> = makeExtractFunctionChanges(doc, doc.uri, params, text, newParams, returnSpec, doc.positionAt(lastToken.pos.max + 1));
+			if (action == null)
+				return [];
+			return [action];
+		} catch (e:Any) {}
+		return [];
+	}
+
+	function makeExtractFunctionChanges(doc:TextDocument, uri:DocumentUri, params:CodeActionParams, text:String, newParams:Array<NewFunctionParameter>,
```

Comment:
Eventually this also needs to take function-level type paramters into account... (the first time I tried it, the resulting function had compilation errors because of this).

Reply:
that and variables "inherited" via `for` or `case`... probably hard to come up with a correct type hint for these